python/cudf/cudf/io/parquet.py

```diff
@@ -561,6 +567,116 @@ def read_parquet(
+
+
+def _normalize_filters(filters: list | None) -> List[List[tuple]] | None:
+    # Utility to normalize and validate the `filters`
+    # argument to `read_parquet`
+    if filters is not None:
+        msg = (
+            f"filters must be None, or non-empty List[Tuple] "
+            f"or List[List[Tuple]]. Got {filters}"
+        )
+        if not filters or not isinstance(filters, list):
+            raise TypeError(msg)
+
+        def _validate_predicate(item):
+            if not isinstance(item, tuple) or len(item) != 3:
+                raise TypeError(
+                    f"Predicate must be Tuple[str, str, Any], "
+                    f"got {predicate}."
+                )
+
+        filters = filters if isinstance(filters[0], list) else [filters]
+        for conjunction in filters:
+            if not conjunction or not isinstance(conjunction, list):
+                raise TypeError(msg)
+            for predicate in conjunction:
+                _validate_predicate(predicate)
+
+    return filters
+
+
+def _handle_in(column, value, *, negate):
+    if not isinstance(value, (list, set, tuple)):
+        raise TypeError(
+            "Value of 'in' or 'not in' filter must be a list, set, or tuple."
+        )
+    if negate:
+        return reduce(operator.and_, (operator.ne(column, v) for v in value))
+    else:
+        return reduce(operator.or_, (operator.eq(column, v) for v in value))
+
+
+def _handle_is(column, value, *, negate):
+    if value not in {np.nan, None}:
+        raise TypeError(
+            "Value of 'is' or 'is not' filter must be np.nan or None."
+        )
+    return ~column.isna() if negate else column.isna()
+
+
+def _apply_post_filters(df: cudf.DataFrame, filters: List[List[tuple]] | None):
+    # Apply DNF filters to an in-memory DataFrame
+    #
+    # Disjunctive normal form (DNF) means that the inner-most
+    # tuple describes a single column predicate. These inner
+    # predicates are combined with an AND conjunction into a
+    # larger predicate. The outer-most list then combines all
+    # of the combined filters with an OR disjunction.
+
+    if not filters:
+        # No filters to apply
+        return df
+
+    handlers: Dict[str, Callable] = {
+        "==": operator.eq,
+        "!=": operator.ne,
+        "<": operator.lt,
+        "<=": operator.le,
+        ">": operator.gt,
+        ">=": operator.ge,
+        "in": partial(_handle_in, negate=False),
+        "not in": partial(_handle_in, negate=True),
+        "is": partial(_handle_is, negate=False),
+        "is not": partial(_handle_is, negate=True),
+    }
+
+    # Can re-set the index before returning if we filter
+    # out rows from a DataFrame with a default RangeIndex
+    # (to reduce memory usage)
+    reset_index = (
+        isinstance(df.index, cudf.RangeIndex)
+        and df.index.name is None
+        and df.index.start == 0
+        and df.index.step == 1
+    )
+
+    try:
+        selection: cudf.Series = reduce(
```

Comment:
Use any/all here.

Reply:
That doesn't work, `any/all` want to produce a single boolean, this is producing a boolean _mask_.

Reply:
I made an error in my comment, discussed here: https://github.com/rapidsai/cudf/pull/13334#discussion_r1194493655